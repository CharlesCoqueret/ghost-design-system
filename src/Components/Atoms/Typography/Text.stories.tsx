import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import mdx from './Typography.mdx';
import TextComponent, { ITextProps, TextTypeEnum } from './Text';

export default {
  title: 'Atom/Typography',
  component: TextComponent,
  parameters: {
    actions: { argTypesRegex: '^on.*' },
    controls: { sort: 'requiredFirst' },
    docs: {
      page: mdx,
    },
  },
  argTypes: {
    type: {
      options: Object.values(TextTypeEnum),
      control: { type: 'select' },
    },
  },
} as ComponentMeta<typeof TextComponent>;

const Template: ComponentStory<typeof TextComponent> = (arg: ITextProps) => {
  return (
    <TextComponent {...arg}>
      Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam rerum nihil
      obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate sed laborum
      explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque quo illum quas
      aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam odio tempora. Quo
      numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel consectetur enim
      aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure velit. Sed impedit eaque 33
      natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa illum ea consectetur perspiciatis ex
      veritatis dolorem id velit sequi qui maiores asperiores!
    </TextComponent>
  );
};

export const Text = Template.bind({});
Text.args = { ellipsis: true };
