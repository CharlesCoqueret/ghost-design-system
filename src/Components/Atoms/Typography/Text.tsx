import React, { CSSProperties, PropsWithChildren, ReactElement } from 'react';
import classNames from 'classnames';

import styles from './Typography.module.scss';

export enum TextTypeEnum {
  BODY = 'body',
  DISABLED = 'disabled',
  ERROR = 'error',
  HELPER = 'helper',
  HIGHLIGHTED = 'highlighted',
  LABEL = 'label',
  PLACEHOLDER = 'placeholder',
  TINY = 'tiny',
}

export interface ITextProps {
  /** custom classname (optional, default: undefined) */
  className?: string;
  /** For test purpose only */
  dataTestId?: string;
  /** Ellipse text when it overflows, or wrap (optional, default: false) */
  ellipsis?: boolean;
  /** Additional style (optional, default: undefined) */
  style?: CSSProperties;
  /** Type of text (optional, default: TextTypeEnum.BODY) */
  type?: TextTypeEnum | Array<TextTypeEnum>;
}

const Text = (props: PropsWithChildren<ITextProps>): ReactElement => {
  const { children, className, dataTestId, ellipsis, style, type } = props;

  return (
    <span
      className={classNames(styles.typography, className, {
        [styles.ellipsis]: ellipsis,
        [styles.error]:
          type && (type === TextTypeEnum.ERROR || (Array.isArray(type) && type.includes(TextTypeEnum.PLACEHOLDER))),
        [styles.disabled]:
          type && (type === TextTypeEnum.DISABLED || (Array.isArray(type) && type.includes(TextTypeEnum.DISABLED))),
        [styles.helper]:
          type && (type === TextTypeEnum.HELPER || (Array.isArray(type) && type.includes(TextTypeEnum.HELPER))),
        [styles.highlighted]:
          type &&
          (type === TextTypeEnum.HIGHLIGHTED || (Array.isArray(type) && type.includes(TextTypeEnum.HIGHLIGHTED))),
        [styles.label]:
          type && (type === TextTypeEnum.LABEL || (Array.isArray(type) && type.includes(TextTypeEnum.TINY))),
        [styles.placeholder]:
          type &&
          (type === TextTypeEnum.PLACEHOLDER || (Array.isArray(type) && type.includes(TextTypeEnum.PLACEHOLDER))),
        [styles.tiny]:
          type && (type === TextTypeEnum.TINY || (Array.isArray(type) && type.includes(TextTypeEnum.TINY))),
      })}
      data-testid={dataTestId}
      style={style}
      title={children && ellipsis && typeof children === 'string' ? children : undefined}>
      {children}
    </span>
  );
};

Text.defaultProps = {
  ellipsis: false,
  onClick: undefined,
  style: undefined,
  type: TextTypeEnum.BODY,
};

export default Text;
