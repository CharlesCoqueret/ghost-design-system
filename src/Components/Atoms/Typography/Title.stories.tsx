import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import mdx from './Typography.mdx';
import TitleComponent, { ITitleProps } from './Title';

export default {
  title: 'Atom/Typography',
  component: TitleComponent,
  parameters: {
    actions: { argTypesRegex: '^on.*' },
    controls: { sort: 'requiredFirst' },
    docs: {
      page: mdx,
    },
  },
} as ComponentMeta<typeof TitleComponent>;

const Template: ComponentStory<typeof TitleComponent> = (arg: ITitleProps) => {
  return (
    <TitleComponent {...arg}>
      Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam rerum nihil
      obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate sed laborum
      explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque quo illum quas
      aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam odio tempora. Quo
      numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel consectetur enim
      aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure velit. Sed impedit eaque 33
      natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa illum ea consectetur perspiciatis ex
      veritatis dolorem id velit sequi qui maiores asperiores!
    </TitleComponent>
  );
};

export const Title = Template.bind({});
Title.args = { ellipsis: true };
