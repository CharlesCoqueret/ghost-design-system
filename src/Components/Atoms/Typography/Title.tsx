import React, { CSSProperties, PropsWithChildren, ReactElement } from 'react';
import classNames from 'classnames';

import styles from './Typography.module.scss';

export interface ITitleProps {
  /** Custom classname (optional, default: undefined) */
  className?: string;
  /** For test purpose only */
  dataTestId?: string;
  /** Ellipse text when it overflows, or wrap (optional, default: false) */
  ellipsis?: boolean;
  /** Header level (optional, default: 2) */
  level?: 1 | 2 | 3;
  /** Additional style (optional, default: undefined) */
  style?: CSSProperties;
}

const Title = (props: PropsWithChildren<ITitleProps>): ReactElement => {
  const { children, className, dataTestId, ellipsis, level, style } = props;

  const HeaderTag = (level && [1, 2, 3].includes(level) ? `h${level}` : 'h2') as keyof JSX.IntrinsicElements;

  return (
    <HeaderTag
      className={classNames(styles.typography, { [styles.ellipsis]: ellipsis }, styles[HeaderTag], className)}
      data-testid={dataTestId}
      style={style}
      title={children && ellipsis && typeof children === 'string' ? children : undefined}>
      {children}
    </HeaderTag>
  );
};

Title.defaultProps = {
  className: undefined,
  ellipsis: false,
  level: 2,
  style: undefined,
};

export default Title;
