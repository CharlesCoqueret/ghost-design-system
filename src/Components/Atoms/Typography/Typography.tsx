import React, { CSSProperties, PropsWithChildren, ReactElement } from 'react';
import classNames from 'classnames';

export interface ITitleProps {
  /** custom classname */
  className?: string;
  /** Ellipse text when it overflows, or wrap (optional, default: false) */
  ellipsis?: boolean;
  /** Header level */
  level: 1 | 2 | 3;
  /** Additional style (optional, default: undefined) */
  style?: CSSProperties;
  /** Click handler (options, default: undefined) */
  onClick?: () => void;
}

const Title = (props: PropsWithChildren<ITitleProps>): ReactElement => {
  const { children, className, ellipsis, level, onClick, style } = props;

  const HeaderTag = `h${level || 3}` as keyof JSX.IntrinsicElements;

  return (
    <HeaderTag
      className={classNames('gds-typography', { ellipsis: ellipsis }, className)}
      onClick={onClick}
      style={style}
      title={children && ellipsis && typeof children === 'string' ? children : undefined}>
      {children}
    </HeaderTag>
  );
};
Title.defaultProps = {
  className: undefined,
  ellipsis: false,
  style: undefined,
};

export enum TextTypeEnum {
  BODY = 'body',
  ERROR = 'error',
  DISABLED = 'disabled',
  HELPER = 'helper',
  HIGHLIGHTED = 'highlighted',
  LABEL = 'label',
  PLACEHOLDER = 'placeholder',
  TINY = 'tiny',
}

export interface ITextProps {
  /** custom classname */
  className?: string;
  /** Ellipse text when it overflows, or wrap (optional, default: false) */
  ellipsis?: boolean;
  /** Click handler (options, default: undefined) */
  onClick?: () => void;
  /** Additional style */
  style?: CSSProperties;
  /** Type of text (optional, default: TextTypeEnum.BODY)*/
  type?: TextTypeEnum | Array<TextTypeEnum | undefined>;
}

const Text = (props: PropsWithChildren<ITextProps>): ReactElement => {
  const { children, className, ellipsis, onClick, style, type } = props;

  const isLink = onClick !== undefined;

  return (
    <span
      className={classNames('gds-typography', className, {
        ellipsis: ellipsis,
        error: type === TextTypeEnum.ERROR || (Array.isArray(type) && type?.includes(TextTypeEnum.PLACEHOLDER)),
        disabled: type === TextTypeEnum.DISABLED || (Array.isArray(type) && type?.includes(TextTypeEnum.DISABLED)),
        helper: type === TextTypeEnum.HELPER || (Array.isArray(type) && type?.includes(TextTypeEnum.HELPER)),
        highlighted:
          type === TextTypeEnum.HIGHLIGHTED || (Array.isArray(type) && type?.includes(TextTypeEnum.HIGHLIGHTED)),
        label: type === TextTypeEnum.LABEL || (Array.isArray(type) && type?.includes(TextTypeEnum.TINY)),
        link: isLink,
        placeholder:
          type === TextTypeEnum.PLACEHOLDER || (Array.isArray(type) && type?.includes(TextTypeEnum.PLACEHOLDER)),
        tiny: type === TextTypeEnum.TINY || (Array.isArray(type) && type?.includes(TextTypeEnum.TINY)),
      })}
      style={style}
      title={children && ellipsis && typeof children === 'string' ? children : undefined}>
      {children}
    </span>
  );
};

Text.defaultProps = {
  ellipsis: false,
  onClick: undefined,
  style: undefined,
  type: TextTypeEnum.BODY,
};

const Typography = {
  Title: Title,
  Text: Text,
};

export { Title, Text };
export default Typography;
