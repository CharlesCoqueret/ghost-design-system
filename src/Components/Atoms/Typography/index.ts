export { default as Typography, TextTypeEnum } from './Typography';
