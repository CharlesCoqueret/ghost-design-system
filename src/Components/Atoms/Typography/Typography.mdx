import { Canvas, ArgsTable, Story, Typeset } from '@storybook/addon-docs';

import Text from './Text';
import Title from './Title';

# Typography

export const typography = {
  type: {
    primary: '"Montserrat", sans-serif',
  },
  weight: {
    regular: '400',
    bold: '600',
    extrabold: '700',
  },
  size: {
    tiny: 10,
    small: 12,
    default: 14,
    h3: 16,
    h2: 18,
    h1: 22,
  },
};

export const SampleText =
  'Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam ' +
  'rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate ' +
  'sed laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque ' +
  'quo illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam ' +
  'odio tempora. Quo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe';

- **Font:** Montserrat

- **Weights:** 400(regular), 600(bold), 700(extrabold)

<Typeset
  fontSizes={[
    Number(typography.size.tiny),
    Number(typography.size.small),
    Number(typography.size.default),
    Number(typography.size.h3),
    Number(typography.size.h2),
    Number(typography.size.h1),
  ]}
  fontWeight={typography.weight.regular}
  sampleText={SampleText}
  fontFamily={typography.type.primary}
/>

## Title

<Canvas withSource='open'>
  <Story id='atom-typography--title' />
</Canvas>

<ArgsTable of={Title} sort='requiredFirst' exclude={['dataTestId']} />

## Text

<Canvas withSource='open'>
  <Story id='atom-typography--text' />
</Canvas>

<ArgsTable of={Text} sort='requiredFirst' exclude={['dataTestId']} />
