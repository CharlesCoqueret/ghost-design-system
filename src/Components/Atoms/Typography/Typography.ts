import Title from './Title';
import Text from './Text';

const Typography = {
  Title: Title,
  Text: Text,
};

export default Typography;
