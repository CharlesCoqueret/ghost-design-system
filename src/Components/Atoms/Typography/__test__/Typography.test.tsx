import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Typography, { TextTypeEnum } from '../Typography';

describe('Typography.Title Component', () => {
  it('Typography.Title renders', async () => {
    const { container } = render(
      <Typography.Title level={2} className='CLASSNAME' ellipsis style={{ height: '200px' }}>
        TITLE
      </Typography.Title>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Typography.Title ellipsis with no text', async () => {
    const { container } = render(
      <Typography.Title ellipsis>
        <div />
      </Typography.Title>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Typography.Title handles click', async () => {
    const onClickMock = jest.fn();

    const { container } = render(
      <Typography.Title dataTestId={'DATA-TEST-ID'} level={1} onClick={onClickMock}>
        TITLE
      </Typography.Title>,
    );
    expect(container).toMatchSnapshot();

    const title = await screen.findByTestId('DATA-TEST-ID');

    userEvent.click(title);

    expect(onClickMock).toBeCalledTimes(1);
  });
});

describe('Typography.Text Component', () => {
  it('Typography.Text renders with no types', async () => {
    const { container } = render(
      <Typography.Text className='CLASSNAME' ellipsis style={{ height: '200px' }}>
        TEXT
      </Typography.Text>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Typography.Text renders with all types', async () => {
    const { container } = render(
      <Typography.Text
        type={[
          TextTypeEnum.BODY,
          TextTypeEnum.DISABLED,
          TextTypeEnum.ERROR,
          TextTypeEnum.HELPER,
          TextTypeEnum.HIGHLIGHTED,
          TextTypeEnum.LABEL,
          TextTypeEnum.PLACEHOLDER,
          TextTypeEnum.TINY,
        ]}>
        TEXT
      </Typography.Text>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Typography.Text renders with all types', async () => {
    const { container: containerBody } = render(<Typography.Text type={TextTypeEnum.BODY}>TEXT</Typography.Text>);
    expect(containerBody).toMatchSnapshot();

    const { container: containerDisabled } = render(
      <Typography.Text type={TextTypeEnum.DISABLED}>TEXT</Typography.Text>,
    );
    expect(containerDisabled).toMatchSnapshot();

    const { container: containerError } = render(<Typography.Text type={TextTypeEnum.ERROR}>TEXT</Typography.Text>);
    expect(containerError).toMatchSnapshot();

    const { container: containerHelper } = render(<Typography.Text type={TextTypeEnum.HELPER}>TEXT</Typography.Text>);
    expect(containerHelper).toMatchSnapshot();

    const { container: containerHightlighted } = render(
      <Typography.Text type={TextTypeEnum.HIGHLIGHTED}>TEXT</Typography.Text>,
    );
    expect(containerHightlighted).toMatchSnapshot();

    const { container: containerLabel } = render(<Typography.Text type={TextTypeEnum.LABEL}>TEXT</Typography.Text>);
    expect(containerLabel).toMatchSnapshot();

    const { container: containerPlaceholder } = render(
      <Typography.Text type={TextTypeEnum.PLACEHOLDER}>TEXT</Typography.Text>,
    );
    expect(containerPlaceholder).toMatchSnapshot();

    const { container: containerTiny } = render(<Typography.Text type={TextTypeEnum.TINY}>TEXT</Typography.Text>);
    expect(containerTiny).toMatchSnapshot();
  });

  it('Typography.Text ellipsis with no text', async () => {
    const { container } = render(
      <Typography.Text ellipsis>
        <div />
      </Typography.Text>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Typography.Text handles click', async () => {
    const onClickMock = jest.fn();

    const { container } = render(
      <Typography.Text dataTestId={'DATA-TEST-ID'} onClick={onClickMock}>
        TEXT
      </Typography.Text>,
    );
    expect(container).toMatchSnapshot();

    const text = await screen.findByTestId('DATA-TEST-ID');

    userEvent.click(text);

    expect(onClickMock).toBeCalledTimes(1);
  });
});
