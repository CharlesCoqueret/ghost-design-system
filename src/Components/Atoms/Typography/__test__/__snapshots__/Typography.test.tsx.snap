// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography.Text Component Typography.Text ellipsis with no text 1`] = `
<div>
  <span
    class="typography ellipsis"
  >
    <div />
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text handles click 1`] = `
<div>
  <span
    class="typography link"
    data-testid="DATA-TEST-ID"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 1`] = `
<div>
  <span
    class="typography error disabled helper highlighted label placeholder tiny"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 2`] = `
<div>
  <span
    class="typography"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 3`] = `
<div>
  <span
    class="typography disabled"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 4`] = `
<div>
  <span
    class="typography error"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 5`] = `
<div>
  <span
    class="typography helper"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 6`] = `
<div>
  <span
    class="typography highlighted"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 7`] = `
<div>
  <span
    class="typography label"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 8`] = `
<div>
  <span
    class="typography placeholder"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with all types 9`] = `
<div>
  <span
    class="typography tiny"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Text Component Typography.Text renders with no types 1`] = `
<div>
  <span
    class="typography CLASSNAME ellipsis"
    style="height: 200px;"
    title="TEXT"
  >
    TEXT
  </span>
</div>
`;

exports[`Typography.Title Component Typography.Title ellipsis with no text 1`] = `
<div>
  <h3
    class="typography ellipsis h3"
  >
    <div />
  </h3>
</div>
`;

exports[`Typography.Title Component Typography.Title handles click 1`] = `
<div>
  <h1
    class="typography h1"
    data-testid="DATA-TEST-ID"
  >
    TITLE
  </h1>
</div>
`;

exports[`Typography.Title Component Typography.Title renders 1`] = `
<div>
  <h2
    class="typography ellipsis h2 CLASSNAME"
    style="height: 200px;"
    title="TITLE"
  >
    TITLE
  </h2>
</div>
`;
